crates/compiler/derive/src/inspect.rs

```diff
@@ -0,0 +1,1096 @@
+                guard: None,
+                redundant: RedundantMark::known_non_redundant(),
+            }
+        })
+        .collect::<Vec<_>>();
+
+    // when tag is
+    //     A v1 v2 -> Inspect.tag "A" [ Inspect.toInspector v1, Inspect.toInspector v2 ]
+    //     B v3 -> Inspect.tag "B" [ Inspect.toInspector v3 ]
+    let when_branches = When {
+        loc_cond: Box::new(Loc::at_zero(Var(tag_sym, tag_union_var))),
+        cond_var: tag_union_var,
+        expr_var: whole_tag_inspectors_var,
+        region: Region::zero(),
+        branches,
+        branches_cond_var: tag_union_var,
+        exhaustive: ExhaustiveMark::known_exhaustive(),
+    };
+
+    // Inspect.custom \fmt -> Inspect.apply (when ..) fmt
+    let (body, this_inspector_var) = wrap_in_inspect_custom(
+        env,
+        when_branches,
+        whole_tag_inspectors_var,
+        tag_sym,
+        tag_union_var,
+    );
+
+    // Create fn_var for ambient capture; we fix it up below.
+    let fn_var = synth_var(env.subs, Content::Error);
+
+    // -[fn_name]->
+    let fn_name_labels = UnionLambdas::insert_into_subs(env.subs, once((fn_name, vec![])));
+    let fn_clos_var = synth_var(
+        env.subs,
+        Content::LambdaSet(LambdaSet {
+            solved: fn_name_labels,
+            recursion_var: OptVariable::NONE,
+            unspecialized: SubsSlice::default(),
+            ambient_function: fn_var,
+        }),
+    );
+    // tag_union_var -[fn_name]-> whole_tag_inspectors_var
+    let tag_union_var_slice = SubsSlice::insert_into_subs(env.subs, once(tag_union_var));
+    env.subs.set_content(
+        fn_var,
+        Content::Structure(FlatType::Func(
+            tag_union_var_slice,
+            fn_clos_var,
+            this_inspector_var,
+        )),
+    );
+
+    // \tag ->
+    //   Inspect.custom \fmt -> Inspect.apply (
+    //     when tag is
+    //        A v1 v2 -> Inspect.tag "A" [ Inspect.toInspector v1, Inspect.toInspector v2 ]
+    //        B v3 -> Inspect.tag "B" [ Inspect.toInspector v3 ])
+    //     fmt
+    let clos = Closure(ClosureData {
+        function_type: fn_var,
+        closure_type: fn_clos_var,
+        return_type: this_inspector_var,
+        name: fn_name,
+        captured_symbols: vec![],
+        recursive: Recursive::NotRecursive,
+        arguments: vec![(
+            tag_union_var,
+            AnnotatedMark::known_exhaustive(),
+            Loc::at_zero(Pattern::Identifier(tag_sym)),
+        )],
+        loc_body: Box::new(Loc::at_zero(body)),
+    });
+
+    (clos, fn_var)
+}
+
+/// Lift `inspector` to `Inspect.custom \fmt -> Inspect.apply inspector fmt`
+///
+/// TODO: currently it appears that just `inspector` is not isomorphic to the lift, on the
+/// monomorphization level, even though we would think it is. In particular, unspecialized lambda
+/// sets fail to resolve when we use the non-lifted version.
+/// More investigation is needed to figure out why.
```

Comment:
We can get rid of this comment, we know why this is.

Reply:
Done